File: backend/src/main/java/com/yourcompany/onlineexam/service/QuestionService.java
package com.yourcompany.onlineexam.service;

import com.google.api.core.ApiFuture;
import com.google.cloud.firestore.*;
import com.google.firebase.cloud.FirestoreClient;
import com.yourcompany.onlineexam.model.Question;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

@Service
public class QuestionService {
    private static final String COLLECTION_NAME = "questions";

    @Autowired
    private QuestionBankService questionBankService;

    public Question create(Question question) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();
        ApiFuture<DocumentReference> future = db.collection(COLLECTION_NAME).add(question);
        String id = future.get().getId();
        question.setId(id);
        db.collection(COLLECTION_NAME).document(id).set(question);
        try {
            updateStats(question.getQuestionBankId());
        } catch (Exception e) {
            // log lỗi nếu cần
        }
        return question;
    }

    public List<Question> getAll(String questionBankId) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();
        CollectionReference ref = db.collection(COLLECTION_NAME);
        Query query = ref.whereEqualTo("questionBankId", questionBankId);
        ApiFuture<QuerySnapshot> future = query.get();
        List<QueryDocumentSnapshot> docs = future.get().getDocuments();
        List<Question> result = new ArrayList<>();
        for (QueryDocumentSnapshot doc : docs) {
            Question q = doc.toObject(Question.class);
            q.setId(doc.getId());
            result.add(q);
        }
        return result;
    }

    private void updateStats(String questionBankId) throws Exception {
        // Lấy tất cả câu hỏi thuộc bankId
        List<Question> questions = getAll(questionBankId);
        int total = questions.size();
        int easy = (int) questions.stream().filter(q -> "easy".equals(q.getLevel())).count();
        int medium = (int) questions.stream().filter(q -> "medium".equals(q.getLevel())).count();
        int hard = (int) questions.stream().filter(q -> "hard".equals(q.getLevel())).count();
        questionBankService.updateQuestionStats(questionBankId, total, easy, medium, hard);
    }

    public void delete(String id, String questionBankId) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();
        db.collection(COLLECTION_NAME).document(id).delete();
        // Cập nhật lại số lượng câu hỏi cho ngân hàng đề
        try {
            updateStats(questionBankId);
        } catch (Exception e) {
            // log lỗi nếu cần
        }
    }

    public Question update(String id, Question question) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();
        question.setId(id);
        db.collection(COLLECTION_NAME).document(id).set(question);
        // Cập nhật lại số lượng câu hỏi cho ngân hàng đề
        try {
            updateStats(question.getQuestionBankId());
        } catch (Exception e) {
            // log lỗi nếu cần
        }
        return question;
    }

    public Question getById(String id) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();
        DocumentSnapshot doc = db.collection(COLLECTION_NAME).document(id).get().get();
        if (doc.exists()) {
            Question question = doc.toObject(Question.class);
            question.setId(doc.getId());
            return question;
        }
        return null;
    }
} 

File: backend/OnlineExam/Services/QuestionService.cs
using Google.Cloud.Firestore;
using OnlineExam.Models;

namespace OnlineExam.Services;

public class QuestionService
{
    private const string CollectionName = "questions";
    private readonly FirestoreDb _db;
    private readonly QuestionBankService _questionBankService;

    public QuestionService(FirestoreDb db, QuestionBankService questionBankService)
    {
        _db = db;
        _questionBankService = questionBankService;
    }

    public async Task<Question> CreateAsync(Question question)
    {
        var docRef = await _db.Collection(CollectionName).AddAsync(question);
        question.Id = docRef.Id;
        await docRef.SetAsync(question);
        try { await UpdateStatsAsync(question.QuestionBankId!); } catch { /* ignore */ }
        return question;
    }

    public async Task<List<Question>> GetAllAsync(string questionBankId)
    {
        var snapshot = await _db.Collection(CollectionName)
            .WhereEqualTo("questionBankId", questionBankId).GetSnapshotAsync();
        return snapshot.Documents.Select(d =>
        {
            var q = d.ConvertTo<Question>();
            q.Id = d.Id;
            return q;
        }).ToList();
    }

    public async Task DeleteAsync(string id, string questionBankId)
    {
        await _db.Collection(CollectionName).Document(id).DeleteAsync();
        try { await UpdateStatsAsync(questionBankId); } catch { /* ignore */ }
    }

    public async Task<Question> UpdateAsync(string id, Question question)
    {
        question.Id = id;
        await _db.Collection(CollectionName).Document(id).SetAsync(question);
        try { await UpdateStatsAsync(question.QuestionBankId!); } catch { /* ignore */ }
        return question;
    }

    public async Task<Question?> GetByIdAsync(string id)
    {
        var snap = await _db.Collection(CollectionName).Document(id).GetSnapshotAsync();
        if (!snap.Exists) return null;
        var question = snap.ConvertTo<Question>();
        question.Id = snap.Id;
        return question;
    }

    private async Task UpdateStatsAsync(string questionBankId)
    {
        var questions = await GetAllAsync(questionBankId);
        var total = questions.Count;
        var easy = questions.Count(q => q.Level == "easy");
        var medium = questions.Count(q => q.Level == "medium");
        var hard = questions.Count(q => q.Level == "hard");
        await _questionBankService.UpdateQuestionStatsAsync(questionBankId, total, easy, medium, hard);
    }
}
